### PR TITLE
[RELENG-158] Bug 1679162 - Split failed task log parsing into multiple queues and workers

### PR DESCRIPTION
**Pull request by escapewindow**

### CODE REVIEW

File: treeherder/log_parser/tasks.py [hunk 1]
```diff
@@ -20,6 +19,8 @@
 logger = logging.getLogger(__name__)
 
 
+# XXX guessing I'm removing this decorator, then adding this decorator to the
+# parser tasks?
```
Comment (camd) [inline treeherder/log_parser/tasks.py:23]: Yeah, exactly.  So add it to `store_failure_lines` and the other one.  I didn't think this through 100%, so some massaging may be necessary there.

File: treeherder/etl/jobs.py [hunk 2]
```diff
@@ -335,12 +335,17 @@ def _schedule_log_parsing(job, job_logs, result):
...
+            queue = "log_parser_fail_json"
+            priority = "failures"
+        else:
+            queue = "log_parser_fail_raw"
+            priority = "failures"
     else:
         queue = 'log_parser'
         priority = "normal"
 
+    # XXX I think I'll change this to turn parse_logs into non-Task?
```
Comment (camd) [inline treeherder/etl/jobs.py:348]: Right, then the async tasks will get assigned to queues one step lower as you mentioned in `tasks.py`

File: treeherder/etl/jobs.py [hunk 3]
```diff
@@ -335,13 +336,22 @@ def _schedule_log_parsing(job, job_logs, result):
...
+            queue += "_sheriffed"
+        else:
+            queue += "_unsheriffed"
     else:
         queue = 'log_parser'
         priority = "normal"
 
-    parse_logs.apply_async(queue=queue, args=[job.id, job_log_ids, priority])
+    # parse_logs.apply_async(queue=queue, args=[job.id, job_log_ids, priority])
+    parse_logs(queue, job.id, job.log_ids, priority)
```
Comment (camd) [inline treeherder/etl/jobs.py:354]: `job.log_ids` should still be `job_log_ids`

Comment (escapewindow) [reply treeherder/etl/jobs.py]: Thanks! fixed.

**Comment by escapewindow** [comment]
Ok, so. As of 0eb5d21, the unit tests are green, and I think it's a better fix than the previous route, which is at https://github.com/escapewindow/treeherder/tree/logparsing-broken , specifically https://github.com/escapewindow/treeherder/commit/55a4c0ed54c29dd5fa70aed3d2cc5ebd782cf920 .

The problem with 55a4c0 is that it looks like a) parse_logs is synchronous, but it waits for the log parsing task to exit, making it synchronous. No? Also, I wasn't able to green up 4 unit tests, which tells me either I broke something, or the unit tests need to be changed to match the new code.

What I like better about 0eb5d21 is we're still running async `parse_logs`, but we call it with a single log per call (and separate queues). And the tests pass locally. I think this should work, but I don't know.

@camd, let me know what you think?

**Comment by escapewindow** [comment]
Green! I'm going to optimistically take this out of draft.

**Comment by camd** [comment]
> Green! I'm going to optimistically take this out of draft.

Nice!  I will review this shortly.  :)

**Comment by camd** [comment]
This is now deployed to Prototype and looks like it's working well so far.  The only thing is that the older `log_parser_fail` queue had two orphaned messages because we stopped processing it in favor of the new queues.  I totally forgot about that.  We'll need to process the old queues (add them to the worker in the `-Q` section) temporarily until the queue is drained, then we can remove it in a follow-up PR.

**Comment by escapewindow** [comment]
> This is now deployed to Prototype and looks like it's working well so far. The only thing is that the older `log_parser_fail` queue had two orphaned messages because we stopped processing it in favor of the new queues. I totally forgot about that. We'll need to process the old queues (add them to the worker in the `-Q` section) temporarily until the queue is drained, then we can remove it in a follow-up PR.

Makes sense. I pushed a fix. I *think* this will work -- we'll parse both the json and raw from the old queue in this worker, but hopefully we'll have a limited number of old messages we'll have to parse.

File: treeherder/log_parser/tasks.py [hunk 4]
```diff
@@ -77,23 +75,6 @@ def parse_logs(job_id, job_log_ids, priority):
     if first_exception:
         raise first_exception
 
-    if "errorsummary_json" in completed_names and "live_backing_log" in completed_names:
```
Comment (camd) [inline treeherder/log_parser/tasks.py:80]: This removal will leave some dead code.  I'll investigate and give some links to code to remove with this.

**Review by camd — CHANGES_REQUESTED** [review]

File: Procfile [hunk 5]
```diff
@@ -44,9 +44,10 @@ worker_store_pulse_data: REMAP_SIGTERM=SIGQUIT newrelic-admin run-program celery
 
 # Handles the log parsing tasks scheduled by `worker_store_pulse_data` as part of job ingestion.
 worker_log_parser: REMAP_SIGTERM=SIGQUIT newrelic-admin run-program celery worker -A treeherder --without-gossip --without-mingle --without-heartbeat -Q log_parser --concurrency=7
-worker_log_parser_fail: REMAP_SIGTERM=SIGQUIT newrelic-admin run-program celery worker -A treeherder --without-gossip --without-mingle --without-heartbeat -Q log_parser_fail --concurrency=1
+worker_log_parser_fail_sheriffed: REMAP_SIGTERM=SIGQUIT newrelic-admin run-program celery worker -A treeherder --without-gossip --without-mingle --without-heartbeat -Q log_parser_fail_raw_sheriffed --concurrency=1
```
[the remark this answers is not in the document]
Comment (escapewindow) [reply Procfile]: Ah, ok. I can drop 54a7d544703059caa67b2b482d07974fe9b9ec88 then.

[the remark this answers is not in the document]
Comment (camd) [reply Procfile]: Yeah, sorry about that.  I was mid-review when I went to lunch and neglected to publish these comments first.  :). Thanks for pivoting.

File: treeherder/log_parser/tasks.py (hunk 4 above)
Comment (camd) [reply treeherder/log_parser/tasks.py:80]: On second thought, I think I should just bite the bullet and create a separate PR to remove this `autoclassify` and `crossreference` code.  That way it won't gum up your PR and we could more easily resurrect it later if we ever decide to.  Let's leave your changes as-is and I'll remove the dead code in a separate PR.  I'll have a few things to check to make sure we don't break anything.

Comment (escapewindow) [reply treeherder/log_parser/tasks.py:80]: Ok! I think I've addressed all your review comments, then. Should we bake https://github.com/mozilla/treeherder/pull/6891/commits/59700923c150ca197ea9520e0ef1d1b0cba38cdf in prototype?

File: Procfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (escapewindow) [reply Procfile]: No worries, thank you for your help!

File: treeherder/log_parser/tasks.py (hunk 4 above)
Comment (camd) [reply treeherder/log_parser/tasks.py:80]: Yep, I'll push your new code to prototype now.  :)

**Review by camd — APPROVED** [review]
What can I say, man.  You did an awesome job here.  :). Thanks for doing this!  It seems to be running great on prototype.  I'm going to merge it, in just a few, but I'll test our queue draining dance on prototype first.

Thanks!!

**Pull request merged**